### ISSUE TITLE
proposal: `SignableContent` trait

### ISSUE THREAD

**Issue by thedavidmeister** [holochain/holochain-rust#948]
something like `AddressableContent` for signing

assuming https://github.com/holochain/holochain-rust/issues/947

```rust
pub trait SignableContent {
 fn sign(&self, ...) -> Provenance {
  ...
 }
}
```

then `Provenance::verify()` would look like:

```rust
impl Provenance {
  pub fn verify(&self, content: SignableContent) -> Boolean {
    ...
  }
}
```

as @maackle pointed out in chat:

> Isn't all Content Signable?

to which I'd say "yes and no"

all content is "addressable content" but having a trait for `AddressableContent` has been very useful pragmatically working with the rust compiler

some points for `SignableContent`:

- many types are signable in theory but in context of holochain, attempting to sign them is a bug (impossible states)
- individual structs/enums may have an opinion on the specific content/order of bytes to sign
- `SignableContent` facilitates expressive yet strict fn signatures

ideally we can ship a default `sign` implementation so that flagging something as `SignableContent` is as simple as:

```rust
impl SignableContent for Foo {}
```

**Comment by maackle** [comment]
@thedavidmeister If it turns out that all Content is both Addressable and Signable, I think `Content` should actually be a trait, i.e. rename AddressableContent to Content and add `sign()`:

```rust
pub trait Content {
    fn address(&self) -> Address; 
    fn content(&self) -> Content;
    fn sign(&self, ...) -> Provenance;
    fn try_from_content(content: &Content) -> Result<Self, HolochainError>;
}

impl Content for JsonString {
    ...
}

type JsonContent = JsonString;  // to make it a little semantically clearer after find-and-replace s/Content/JsonContent/
```

If it turns out that we want to tease apart Addressability and Signability, then if we have both `AddressableContent` and `SignableContent`, both traits would have a `content()` method and there would be a collision. So in that case it might be better to factor out Content from the two other traits:

```rust
trait Content {
    fn content(&self) -> Content;
    fn try_from_content(content: &Content) -> Result<Self, HolochainError>;
}
trait Addressable {
    fn address(&self) -> Address where Self: Content;
    ...
}
trait Signable {
    fn sign(&self, ...) -> Provenance where Self: Content;
    ...
}
```

Then we can only `impl Signable for Foo` if also `impl Content for Foo`.

**Comment by thedavidmeister** [comment]
RTB link https://realtimeboard.com/app/board/o9J_kyiXmFs=/?moveToWidget=3074457346432940491

**Comment by lucksus** [comment]
I don't think we would sign content directly anyway since this would allow for signature forgery attacks (https://en.wikipedia.org/wiki/Digital_signature_forgery). To prevent such attacks, content is usually first hashed and then that hash is signed. So I don't see the need to add a `Signable` trait. Everything that is `AddressableContent` can also be signed by using the address as the message to sign.

Also, for the design that I have in mind that we need for the Holo alpha interims solution, we have JSON-RPC between core and the keys. So we could not apply Rust's type system here if we stick to that solution. Nor do we want to pass around the full content between different modules (core <-> conductor <-> Holo light light client) which could potentially be big. Instead we want to just pass around the hash/address of content to be signed.

Pulling in @neonphog to have him confirm or correct what I just said.

**Comment by neonphog** [comment]
The libsodium [docs](https://download.libsodium.org/doc/public-key_cryptography/public-key_signatures#notes) are a little unclear, except to say explicitly that [passing the data directly is recommended](https://download.libsodium.org/doc/public-key_cryptography/public-key_signatures#multi-part-messages).

Looking at the [code](https://github.com/jedisct1/libsodium/blob/master/src/libsodium/crypto_sign/ed25519/ref10/sign.c#L81), I * think * the signing algorithm already does a sha512 on the input data.

Which doesn't mean it wouldn't work to prehash on our side as well... just adds another layer of indirection.

**Comment by thedavidmeister** [comment]
> To prevent such attacks, content is usually first hashed and then that hash is signed. So I don't see the need to add a `Signable` trait.

@lucksus neat trick! i didn't know about that

> Everything that is `AddressableContent` can also be signed by using the address as the message to sign.

unfortunately not... for the same reason you highlighted

`address` is explicitly NOT guaranteed to be a hash, that's why `AddressableContent` exists as a trait

> So I don't see the need to add a `Signable` trait.

OK, happy to use `AddressableContent` if it does everything we need already

> which could potentially be big. Instead we want to just pass around the hash/address of content to be signed.

yes, would it make sense to hash the address and pass that around? would mean double hashing but also ensure we don't expose edge-case issues through `.address()` implementations

> Which doesn't mean it wouldn't work to prehash on our side as well...

@neonphog yeah... what you're saying makes sense, but also if signing is provided by the conductor then i'd feel more comfortable being defensive here (e.g. for community maintained conductors that might not realise this is needed).
